rust/src/builtins/apply_live.rs

```diff
@@ -94,8 +95,39 @@ pub(crate) fn applylive_finish(sysroot: &crate::ffi::OstreeSysroot) -> CxxResult
     if pkgdiff.n_removed() == 0 && pkgdiff.n_modified() == 0 {
         crate::ffi::output_message("Successfully updated running filesystem tree.");
     } else {
-        crate::ffi::output_message(
-            "Successfully updated running filesystem tree; some services may need to be restarted.",
+        let lib_diff = ostree_ext::diff::diff(
+            repo,
+            &booted_commit,
+            &live_state.commit.as_str(),
+            Some("/usr/lib/systemd/system"),
+        )?;
+
+        let etc_diff = ostree_ext::diff::diff(
+            repo,
+            &booted_commit,
+            &live_state.commit.as_str(),
+            Some("/usr/etc/systemd/system"),
+        )?;
+
+        if !lib_diff.changed_files.is_empty()
+            || !etc_diff.changed_files.is_empty()
+            || !lib_diff.added_files.is_empty()
+            || !etc_diff.added_files.is_empty()
+        {
+            let output = Command::new("/usr/bin/systemctl")
+                .arg("daemon-reload")
+                .status()
+                .expect("Failed to reload systemd manager configuration");
+            assert!(output.success());
+        }
+        let changed: Vec<String> = lib_diff
+            .changed_files
+            .union(&etc_diff.changed_files)
+            .cloned()
```

Comment:
This is completely fine as is, I would just note that we could likely avoid the allocation/clone by doing instead `.map(|s| s.as_str())`, producing a `Vec<&str>`.